loopy/tools.py

```diff
@@ -862,4 +865,40 @@ def t_unit_to_python(t_unit, var_name="t_unit",
     else:
         return python_code
 
+
+def memoize_on_disk(func, key_builder_t=LoopyKeyBuilder):
+    from loopy.version import DATA_MODEL_VERSION
+    from functools import wraps
+    from pytools.persistent_dict import WriteOncePersistentDict
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        from loopy import CACHING_ENABLED
+
+        if (not CACHING_ENABLED
+                or kwargs.pop("_no_memoize_on_disk", False)):
+            return func(*args, **kwargs)
+
+        transform_cache = WriteOncePersistentDict(
+            (f"loopy-memoize-cache-{key_builder_t.__qualname__}"
+             f"-{key_builder_t.__name__}-v0-{DATA_MODEL_VERSION}"),
+            key_builder=key_builder_t())
+        cache_key = (args, kwargs)
+
+        try:
+            result = transform_cache[cache_key]
+            logger.debug(f"Function {func.__name__} returned from"
+                         " memoized result on disk.")
+            return result
+        except KeyError:
+            logger.debug(f"Function {func.__name__} not present"
+                         " on disk.")
+
+            with ProcessLogger(logger, f"{func.__name__}"):
```

Comment:
I get that this function is not specific to things that receive kernels or programs, but we are inside of loopy here, and the dominant use case is that. Could you be convinced to at least check if there's a `TranslationUnit` or `LoopKernel` in the args, and if so, show the name? I do think it's important for the user experience.

Reply:
Yep agreed, done.